pcap_linux.go

```diff
@@ -108,15 +110,40 @@ func (h *Handle) ReadPacketData() (data []byte, ci gopacket.CaptureInfo, err err
 
 func (h *Handle) readPacketDataSyscall() (data []byte, ci gopacket.CaptureInfo, err error) {
 	b := make([]byte, h.snaplen)
-	read, _, err := syscall.Recvfrom(h.fd, b, 0)
+	oob := make([]byte, syscall.CmsgSpace(tpacketAuxdataSize))
+	n, _, _, _, err := syscall.Recvmsg(h.fd, b, oob, 0)
 	if err != nil {
-		return nil, ci, fmt.Errorf("error reading: %v", err)
+		return nil, ci, fmt.Errorf("error reading packets: %w", err)
+	}
+
+	var auxData syscall.TpacketAuxdata
+	cmsgs, err := syscall.ParseSocketControlMessage(oob)
+	if err != nil {
+		return nil, ci, fmt.Errorf("error reading socket control messages: %w", err)
+	}
+	for _, cmsg := range cmsgs {
+		if cmsg.Header.Level == syscall.SOL_PACKET && cmsg.Header.Type == syscall.PACKET_AUXDATA && cmsg.Header.Len >= tpacketAuxdataSize {
+			auxData.Vlan_tci = binary.BigEndian.Uint16(cmsg.Data[len(cmsg.Data)-5 : len(cmsg.Data)-3])
+			auxData.Vlan_tpid = binary.BigEndian.Uint16(cmsg.Data[len(cmsg.Data)-3:])
+			break
+		}
+	}
+	if auxData.Vlan_tci != 0 {
+		if auxData.Vlan_tpid == 0 {
+			auxData.Vlan_tpid = binary.BigEndian.Uint16(b[12:14])
+			binary.BigEndian.PutUint16(b[12:14], 0x8100) // set ethernet frame type to VLAN
+		}
+		aux := make([]byte, 4)
+		binary.BigEndian.PutUint16(aux[:2], auxData.Vlan_tci)
+		binary.BigEndian.PutUint16(aux[2:], auxData.Vlan_tpid)
+		b = append(append(b[:14], aux...), b[14:]...)
+		n = n + 4
 	}
 	// TODO: add CaptureInfo, specifically:
 	//    capture timestamp
 	//    original packet length
 	ci = gopacket.CaptureInfo{
-		CaptureLength:  read,
+		CaptureLength:  n,
```

Comment:
Is this the same thing? Why change it from `read` to `n`?

Reply:
So I originally had `n, noob, _, _, err := syscall.RecvMsg(...)` where `noob` is the bytes read of the cmsg, and `CaptureLength` was the sum of `n` and `noob`. I then realized in this case, it's only ever `n` or `n+4`, I can certainly rename it to `read` and leave this line as is if desired.

Reply:
Nah, I am fine with it changed to `n`, if it doesn't matter